pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,662 @@
+			+ TryInto<<Self::ReserveCurrency as Currency<Self::AccountId>>::Balance>;
+
+		/// Type used for Nonce used in OrderId generation.  Nonce ensures each
+		/// OrderId is unique. Nonce incremented
+		type Nonce: Parameter
+			+ Member
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ EnsureAdd
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen;
+
+		/// Type for currency orders can be made for
+		type TradeableAsset: MultiReservableCurrency<
+			Self::AccountId,
+			Balance = <Self as pallet::Config>::ForeignCurrencyBalance,
+			CurrencyId = Self::AssetCurrencyId,
+		>;
+
+		/// Type for pallet weights
+		type Weights: WeightInfo;
+	}
+	//
+	// Storage and storage types
+	//
+	/// Order Storage item.
+	/// Contains fields relevant to order information
+	#[derive(Clone, Copy, Debug, Encode, Decode, Eq, PartialEq, MaxEncodedLen, TypeInfo)]
+	pub struct Order<OrderId, AccountId, AssetId, ForeignCurrencyBalance> {
+		pub order_id: OrderId,
+		pub placing_account: AccountId,
+		pub asset_in_id: AssetId,
+		pub asset_out_id: AssetId,
+		/// How many tokens of asset in to buy
+		pub buy_amount: ForeignCurrencyBalance,
+		/// Original buy amount, used for tracking amount fulfilled
+		pub initial_buy_amount: ForeignCurrencyBalance,
+		/// How much currency being purchased (asset in) costs with asset sold
+		/// (asset out)
+		pub price: ForeignCurrencyBalance,
+		/// Minimum amount of an order that can be fulfilled
+		/// for partial fulfillment
+		pub min_fullfillment_amount: ForeignCurrencyBalance,
+		/// Maximum amount of outgoing currency that can be sold
+		pub max_sell_amount: ForeignCurrencyBalance,
+	}
+
+	/// Map of Orders to look up orders by their order id.
+	#[pallet::storage]
+	pub type Orders<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		T::Hash,
+		Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::ForeignCurrencyBalance>,
+		ResultQuery<Error<T>::OrderNotFound>,
+	>;
+
+	/// Map of orders for a particular user
+	/// Used to query orders for a particular user using the
+	/// account id of the user as prefix
+	#[pallet::storage]
+	pub type UserOrders<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::AccountId,
+		Twox64Concat,
+		T::Hash,
+		OrderOf<T>,
+		ResultQuery<Error<T>::OrderNotFound>,
+	>;
+
+	/// Stores Nonce for orders placed
+	/// Given that Nonce is to ensure that all orders have a unique ID, we can
+	/// use just one Nonce, which means that we only have one val in storage,
+	/// and we don't have to insert new map values upon a new account/currency
+	/// order creation.
+	#[pallet::storage]
+	pub type NonceStore<T: Config> = StorageValue<_, T::Nonce, ValueQuery>;
+
+	/// Map of Vec containing OrderIds of same asset in/out pairs.
+	/// Allows looking up orders available corresponding pairs.
+	#[pallet::storage]
+	pub type AssetPairOrders<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::AssetCurrencyId,
+		Twox64Concat,
+		T::AssetCurrencyId,
+		BoundedVec<T::Hash, ConstU32<1_000_000>>,
```

Comment:
Good call on the bounded vec size, I'll get that update.

The AssetPairOrders is to provide a quick means of querying storage for orders that match a desired trade.

Reply:
Updated to use pallet const conf type.

Reply:
>The AssetPairOrders is to provide a quick means of querying storage for orders that match a desired trade.

And where is it used? Is it intended to be used by the front-end?

Reply:
Yes, this is primarily for front end right now, it will also be necessary if/when we add fulfilling orders by opposite asset ids with compatible prices.